common/power_button.c

```diff
@@ -115,6 +115,624 @@ static void power_button_init(void)
+ * #define CONFIG_AFTER_G3_STATE_USE_AP_OFF_AS_OFF_FLAG
+ *
+ *==========
+ *== Enables fix for bbram bug in older EC code, where presence of *any* bbram
+ *== reset flags after a jump will discard jump data, and not mark image state
+ *== as jumped to.
+ *==
+ *== Bug is caused by the fact that when EC image is executed after a jump, it
+ *== first restores system reset flags from bbram reset flags via pathway:
+ *==   main()->gpio_pre_init()->system_is_reboot_warm()->
+ *==     system_check_reset_cause()
+ *== Then it overwrites system reset flags with jump data reset flags via
+ *== pathway:
+ *==   main()->system_common_pre_init()
+ *==
+ *== In bugged firmware system_common_pre_init() will only restore jump data if
+ *== condition "system_reset == 0" (system reset flags == 0) evaluates to TRUE.
+ *==
+ *== This creates a risk of causing auto power on if device is off, or risk of
+ *== inducing abnormal state (initializing power button as off) if device is
+ *== running.
+ *==
+ *== Upstream fixed this is newer EC code by dropping the problematic condition:
+ *==   https://github.com/MrChromebox/chrome-ec/commit/e833bdec81b9
+ *== However, this is mainly suitable for fixing RW (jump RO->RW), as fixing RO
+ *== would require overwriting factory RO, which poses a risk of bricking
+ *== device, with no easy way to recover.
+ *==
+ *== Fixing RO (jump RW->RO), without actually changing RO, has two pathways:
+ *==   1) Non-OFF bbram reset flags are NOT set when jumping. Fix will unset
+ *==      any set OFF bbram reset flags. This has the following potential side
+ *==      effects:
+ *==        - If EC_AFTER_G3_STATE_OFF is set, while device is off/running, then
+ *==          losing power before returning to RW will auto power on device when
+ *==          power is restored.
+ *==        - If EC_AFTER_G3_STATE_OFF/EC_AFTER_G3_STATE_PREVIOUS is set, while
+ *==          device is off, then losing power power before returning to RW will
+ *==          auto power on device when power is restored.
+ *==   2) Non-OFF bbram reset flags are set when jumping. This means that jump
+ *==      will be bugged anyway, since we do not want to touch non-OFF bbram
+ *==      reset flags. Fix will adjust OFF bbram reset flags in a way to ensure
+ *==      no abnormal state after jump. This has the following potential side
+ *==      effects:
+ *==        - If EC_AFTER_G3_STATE_OFF is set, while device is running, then
+ *==          losing power before returning to RW will auto power on device when
+ *==          power is restored.
+ *==        - If EC_AFTER_G3_STATE_ON is set, while device is off, then losing
+ *==          power before returning to RW will keep device off when power is
+ *==          restored.
+ *==
+ *== Useless for PUFF, and newer boards, as these should not be affected.
+ *==========
+ * #define CONFIG_AFTER_G3_STATE_SYSJUMP_BBRAM_BUGFIX
+ *
+ *==========
+ *== DEDEDE, and newer boards, do not pass OFF system reset flags via jump data
```

Comment:
how is this relevant for Fizz?

Reply:
I would prefer the cleanest, leanest solution for each branch. Fizz is the only baord I've had any complaints about it not working properly, so let's just fix that. Ideally, RW (or RO->RW) only. KISS